/***************************************************************************
 *   energydlg.cpp                                                         *
 *   Copyright (C) 2014 by Daniel Suson                                    *
 *   suson@calumet.purdue.edu                                              *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 *                                                                         *
 *   In addition, as a special exception, the copyright holders give       *
 *   permission to link the code of this program with any edition of       *
 *   the Qt library by Trolltech AS, Norway (or with modified versions     *
 *   of Qt that use the same license as Qt), and distribute linked         *
 *   combinations including the two.  You must obey the GNU General        *
 *   Public License in all respects for all of the code used other than    *
 *   Qt.  If you modify this file, you may extend this exception to        *
 *   your version of the file, but you are not obligated to do so.  If     *
 *   you do not wish to do so, delete this exception statement from        *
 *   your version.                                                         *
 ***************************************************************************
 * Description:                                                            *
 *                                                                         *
 * This class defines the energy dialog structure associated with          *
 * the Qt4 user interface.                                                 *
 * ----------------------- Change Log ------------------------------------ *
 * Date    Name         Description                                        *
 *                                                                         *
 ***************************************************************************
 * This software evolved from work that was done by Eric Hivon at the      *
 * California Institute of Technology on the MASTER algorithm (Hivon, E.,  *
 * et al., ApJ, vol 567, p 2).  It adapts or interfaces with these         *
 * packages:                                                               *
 *                                                                         *
 * HealPix         (http://www.eso.org/science/healpix/)                   *
 * FTOOLS & FITSIO (http://heasarc.gsfc.nasa.gov/docs/corp/software.html)  *
 *                                                                         *
 * Any publications or presentations that use results generated by this    *
 * program must reference the fact that it was developed by Daniel Suson   *
 ***************************************************************************/
#include "energydlg.h"
#include "ui_energydlg.h"

energyDialog::energyDialog() :
              ui(new Ui::energyDialog) {
  /* set up the user interface first */
  ui->setupUi(this);
  
  /* initialize internal variables*/
  minEnergy = 0.00;
  maxEnergy = 1000000.00;
  dirty = false;
  isConfigured = false;
  
  /* set up signals and slots */
  connect(ui->buttonBox, &QDialogButtonBox::accepted, this, &energyDialog::finalize);
  connect(ui->buttonBox, &QDialogButtonBox::rejected, this, &energyDialog::cancel);
  connect(ui->buttonBox, &QDialogButtonBox::helpRequested, this, &energyDialog::help);
}

energyDialog::~energyDialog() {
  delete ui;
}

void energyDialog::configure(double min, double max) {
  dirty = false;
  ui->minEnergyEdit->insert(QString::number(min));
  ui->maxEnergyEdit->insert(QString::number(max));
  exec();
}

void energyDialog::reset() {
  minEnergy = 0.0;
  maxEnergy = 1000000.0;
  ui->minEnergyEdit->insert(QString::number(minEnergy));
  ui->maxEnergyEdit->insert(QString::number(maxEnergy));
  dirty = false;
  isConfigured = false;
}

void energyDialog::validate() {
  QString title, message;
  bool ok = false;
  
  dirty = false; // assume nothing has changed
  
  /* set energy */
  double oldminE = minEnergy;
  double oldmaxE = maxEnergy;
  minEnergy = ui->minEnergyEdit->text().toDouble(&ok);
  if (!ok) {
    title = QString(tr("Error Converting Minimum Energy"));
    message = QString(tr("Something went wrong converting the minimum energy from a string to a number.\nPlease let the developers know."));
    QMessageBox::critical(this,title,message);
  }
  maxEnergy = ui->maxEnergyEdit->text().toDouble(&ok);
  if (!ok) {
    title = QString(tr("Error Converting Maximum Energy"));
    message = QString(tr("Something went wrong converting the maximum energy from a string to a number.\nPlease let the developers know."));
    QMessageBox::critical(this,title,message);
  }
  if (minEnergy != oldminE || maxEnergy != oldmaxE)
    dirty = true;
  
  isConfigured = true;
  return;
}

void energyDialog::finalize() {
  int err = 0;
  
  validate();
 
  accept();
  Q_EMIT energySelected(minEnergy,maxEnergy);
}

void energyDialog::help() {

}

void energyDialog::cancel() {
  QString title, message;
  QMessageBox::StandardButton reply;
  
  if (dirty) {
    title = QString(tr("Energy range changed"));
    message = QString(tr("You have changed the selected energy range.\nDo you really want to quit?"));
    reply = QMessageBox::warning(this,title,message,QMessageBox::Yes | QMessageBox::No,QMessageBox::No);
    if (reply == QMessageBox::No)
      return;
    dirty = false;
  }
  Q_EMIT energySelectionCancelled();
  close();
}
